Show viewpoints from BCF gallery

oliverguenther opened the pull request:
- [x] Get viewpoint from viewer
- [x] Abstract revit and xeokit services into one using `useFactory`

**TODOS**
- [x] Save viewpoint works, but does not reload work package
- [x] ~~snapshot of saved viewpoint is seemingly empty~~ Caused by incorrectly parsing the entire base64 URL, instead of only the encoded parts
- [x] Bitmaps need to be removed in frontend

wielinde committed:
Improve code climate

wielinde committed:
Fix: @HostListener does not work in a angular service. [ci skip]

wielinde committed:
Add BCF topic UUID and viewpoint UUIDs to WorkPackageRepresenter

wielinde committed:
Rename global object to RevitBridge

wielinde committed:
Add temporary button for showing viewpoint in Revit

wielinde committed:
WIP

wielinde committed:
Fix: import of filter.

wielinde committed:
Fix bcfSnapshotPath

Commit:
Fix: only use one camera attribute and ensure vectors are orthogonal

wielinde committed:
Enrich Revit BCF snapshots with snapshot_type

wielinde committed:
Fix: Adding viewpoint from Revit and saving it in backend

wielinde committed:
Change ngxGallery to allow for showing viewpoints in Revit.

wielinde committed:
Common interface for getting and setting viewpoints

wielinde committed:
Resolve merge conflict

wielinde committed:
Fix: Show View thumbnails of viewpoint snapshots

wielinde committed:
WIP

oliverguenther committed:
Merge remote-tracking branch 'origin/dev' into bim/feature/create-viewpoints-from-work-package

oliverguenther committed:
Fix showing viewpoints in gallery again

oliverguenther committed:
Fix h3 structure of bcf attribute group

oliverguenther committed:
Show viewpoint from BCF gallery

openprojectci commented on frontend/src/app/modules/bim/bcf/bcf-wp-single-view/bcf-wp-single-view.component.ts:
```diff
@@ -2,49 +2,76 @@ import {Component, Input, OnDestroy, OnInit} from "@angular/core";
+
+export type ViewPointOriginal = { uuid:string, snapshot_id:string, snapshot_file_name:string };
+export type ViewPoint = { snapshotId:string, snapshotFileName?:string, snapshotFullPath:string };
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/603cab1b34ef23edd1c7087100cbc18fec4969a9/frontend/src/app/modules/bim/bcf/bcf-wp-single-view/bcf-wp-single-view.component.ts#L15"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

oliverguenther committed:
Move stuff from old bcf module to bim and unify service provider

We can avoid the ModelViewService which is only delegating to revit or
xeokit services by simply providing the correct service

oliverguenther committed:
Fix add button by adding directly to parent component

oliverguenther committed:
Correctly save preliminary viewpoint

oliverguenther committed:
Fix incorrect parsing of base64 encoded snapshot

oliverguenther committed:
Let work package cache associated bcf_issue

and touch it when changing associated viewpoints

oliverguenther committed:
Remove empty bitmaps in xeokit server

oliverguenther committed:
Create bcf issue if not exists

oliverguenther committed:
Use onPush for bcf component

oliverguenther commented:
**Note**: Creating a viewpoint from a newly created work package (without BCF issue) does not yet work, because the issue created always creates a work package. We depend on https://community.openproject.com/wp/32493 for this.

HDinger committed:
Replace inline style by class

oliverguenther committed:
Add spec for showing viewpoint

oliverguenther committed:
Add timestamps to properly cache invalidate and add spec

oliverguenther committed:
Only show bcf when viewer is present

HDinger reviewed (commented):
Looks good 👍 As said, I ignored the style issues for now. 
I found two issues, however at least one is probably the same the tests are finding: that you cannot open the split or the full screen of a WP on the WP page.

The other one happens when you open the split screen of a non-BCF Work Package in the BIM module. There are things missing, and most of the Angular buttons don't work correctly any more.
<img width="1680" alt="Bildschirmfoto 2020-03-10 um 13 14 47" src="https://user-images.githubusercontent.com/7457313/76311804-06811b00-62d2-11ea-8f9e-34f1dfb6c6a7.png">

oliverguenther committed:
Fix specs

oliverguenther commented:
@HDinger I addressed all remarks. the filter spec is still failing but working locally.
